### PR TITLE
docs(pull-requests.md): add pull-requests.md

### PR DESCRIPTION
**Pull request by pon-james**
Adds documentation for expectations on pull requests taking inspiration from https://gist.github.com/mikepea/863f63d6e37281e329f8 and https://github.com/alphagov/styleguides/blob/master/pull-requests.md

### CODE REVIEW

File: pull-requests.md [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+PRs are a Github workflow tool, so it's important to understand that the PR title, summary and eventual discussion are not as trackable as the the commit history. If we ever move away from Github, we'll likely lose this information.
+
+That said however, they are a very useful aid in ensuring that PRs are handled quickly and effectively.
+
+Ensure that your PR title is scannable. People will read through the list of PRs attached to a repo, and must be able to distinguish between them based on title. Include a ticket/issue reference if possible, so the reviewer can get any extra context. Include a reference to the subsystem affected, if this is a large codebase.
+
+It is recommended to use the [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/) commit structure where appropriate. It provides an easy set of rules for creating an explicit commit history; which makes it easier to write automated tools on top of.
+
+
+### Rebase before you make the PR, if needed
```
Comment (zeger-pon) [inline pull-requests.md:44]: I think we should make this a single line with a reference to branching / rebasing guidelines?

File: pull-requests.md [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+You should not rewrite commits that have been pushed unless you:
+
+* are very sure that no-one else will be affected by you rewriting the branch history
+* have an Extremely Good Reason. For example: someone has committed sensitive information (personally identifiable data, passwords and suchlike) and it needs purging from history
+
+### Aim for one succinct commit
+
+In an ideal world, your PR will be one small(ish) commit, doing one thing - in which case your life will be made easier, since the commit message and PR title/summary are equivalent.
+
+If your change contains more work than can be sensibly covered in a single commit though, **do not** try to squash it down into one. Commit history should tell a story, and if that story is long then it may require multiple commits to walk the reviewer through it. Instead, use the "Squash and merge" option to combine all your commits into one a single commit on the target branch.
```
Comment (zeger-pon) [inline pull-requests.md:59]: Same as previous comment

File: pull-requests.md [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+
+
+
+
+Reviewing Pull Requests
+-----
+
+It's a reviewers responsibility to ensure:
+
+* Commit history is excellent
```
Comment (zeger-pon) [inline pull-requests.md:92]: what is excellent? :)

File: pull-requests.md [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+* Good changes are propagated quickly
```
Comment (zeger-pon) [inline pull-requests.md:93]: What is good?

File: pull-requests.md [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+* Code review is performed
```
Comment (zeger-pon) [inline pull-requests.md:94]: Rather rephrase to code review governance is in place

File: pull-requests.md [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+* Anti patterns are flagged
+* Out of scope changes are omitted
+* They understand what is being changed, from the perspective of someone examining the code in the future.
```
Comment (zeger-pon) [inline pull-requests.md:97]: Refer to development guide

File: pull-requests.md [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+
+### Reviewers are the guardians of the commit history
+
+The importance of ensuring a quality commit history cannot be stressed enough. It is the historical context of all of the work that we do, and is vital for understanding the reasons why changes were made in the past. What is obvious now, will not be obvious 2-3 years in the future.
+
+Without a decent commit history, we may as well be storing all our code in files ending yyyy-mm-dd. The commit history of a code base is what allows people to understand **why** a change was made - the when, what, and where are automatically evident.
+
+When looking at a commit message, ask yourself the question - from the perspective of someone looking at this change without any knowledge of the codebase - 'do I understand *why* this change was made?'
```
Comment (zeger-pon) [inline pull-requests.md:105]: Start with this, this is the "why"

File: pull-requests.md [hunk 8]
```diff
@@ -0,0 +1,162 @@
...
+
+
+
+### Keep the flow going
+
+Pull Requests are the fundamental unit of how we progress change. If PRs are getting clogged up in the system, pending review, they are preventing a piece of work from being completed.
```
Comment (zeger-pon) [inline pull-requests.md:111]: Good chapter! There should be a reference to the customer / business here

File: pull-requests.md [hunk 9]
```diff
@@ -0,0 +1,162 @@
...
+As PRs clog up in the system, merges become more difficult, as other features and fixes are applied to the same codebase. This in turn slows them down further, and often completely blocks progress on a given codebase.
+
+There is a balance between flow and ensuring the quality of our PRs. As a reviewer you should make a call as to whether a code quality issue is sufficient enough to block the PR.
+
+Any quality issue that will obviously result in a bug should be fixed.
+
+
+### We are all reviewers
+
+To make sure PRs flow through the system speedily, we must scale the PR review process. It is not sufficient (or fair!) to expect one or two people to review all PRs to our code. For starters, it creates a blocker every time those people are busy.
```
Comment (zeger-pon) [inline pull-requests.md:122]: Need a stronger first line here - a real clear "why"

File: pull-requests.md [hunk 10]
```diff
@@ -0,0 +1,162 @@
...
+
+Hopefully with the above guidelines, we can all start sharing the responsibility of being a reviewer.
+
+NB: With this in mind - if you are the first to comment on a PR, you are that PRs reviewer. If you feel that you can no longer be responsible for the subsequent merge or closure of the PR, then flag this up in the PR conversation, so someone else can take up the role.
+
+There's no reason why multiple people cannot comment on a PR and review it, and this is to be encouraged.
+
+
+### Don't add to the PR yourself.
```
Comment (zeger-pon) [inline pull-requests.md:131]: Nope - if you add to the PR you are no longer a reviewer (which is oke, as long as you are aware of this)

File: pull-requests.md [hunk 11]
```diff
@@ -0,0 +1,162 @@
...
+
+It's sometimes tempting to fix a bug in a PR yourself, or to rework a section to meet coding standards, or just to make a feature better fit your needs.
+
+If you do this, you are no longer the reviewer of the PR. You are a collaborator, and so should not merge the PR.
+
+It is of course possible to find a new reviewer, but generally change will be speedier if you require the original contributor to fix the code themselves. Tell them what you see (the attention point) and what you expect to see (the desired result). Alternatively, if the original PR is not'good enough', raise the changes you'd like to see in your own PR.
+
+### Be supportive
```
Comment (zeger-pon) [inline pull-requests.md:139]: This should be moved to the to top, I think based on Pon (make it fun) it should be something like "a review points out both the challenges and the good parts)

File: pull-requests.md [hunk 12]
```diff
@@ -0,0 +1,162 @@
...
+**Poor quality code can be refactored. A terrible commit lasts forever.**
+
+
+### What constitutes a good PR?
+
+A good quality PR will have the following characteristics:
+
+* It will be a complete piece of work that adds value in some way.
+* It will have a title that reflects the work within, and a summary that helps to understand the context of the change.
+* There will be well written commit messages, with well crafted commits that tell the story of the development of this work.
```
[the remark this answers is not in the document]
Comment (pon-james) [reply pull-requests.md]: Addressed in 09378360c9d5431828344774ac8ab22cdd039a4d

File: pull-requests.md [hunk 13]
```diff
@@ -0,0 +1,162 @@
...
+* Ideally it will be small and easy to understand. Single commit PRs are usually easy to submit, review, and merge.
```
[the remark this answers is not in the document]
Comment (pon-james) [reply pull-requests.md]: Addressed in 09378360c9d5431828344774ac8ab22cdd039a4d

File: pull-requests.md [hunk 14]
```diff
@@ -0,0 +1,162 @@
...
+* The code contained within will meet the [best practices](./README.md#best-practices) wherever possible.
+
+A PR does not end at submission though. A code change is not made until it is merged and used in production.
+
+A good PR should be able to flow through a peer review system easily and quickly.
```
[the remark this answers is not in the document]
Comment (pon-james) [reply pull-requests.md]: Addressed in 09378360c9d5431828344774ac8ab22cdd039a4d

File: pull-requests.md (hunk 9 above)
Comment (pon-james) [reply pull-requests.md:118]: Addressed in 09378360c9d5431828344774ac8ab22cdd039a4d

File: pull-requests.md (hunk 10 above)
Comment (pon-james) [reply pull-requests.md]: Addressed in 09378360c9d5431828344774ac8ab22cdd039a4d

File: pull-requests.md (hunk 11 above)
Comment (pon-james) [reply pull-requests.md:133]: At the top of the section or top of the document? I see a paragraph and title highlighted. It helps me if only the affected lines are highlighted.

File: pull-requests.md [hunk 15]
```diff
@@ -0,0 +1,162 @@
...
+Done well, the commits (and their attached messages) contained within tell a story to people examining the code at a later date. If we are not careful to ensure the quality of these commits, we silently lose this ability.
+
+**Poor quality code can be refactored. A terrible commit lasts forever.**
+
+
+### What constitutes a good PR?
+
+A good quality PR will have the following characteristics:
+
+* It will be a complete piece of work that adds value in some way.
```
[the remark this answers is not in the document]
Comment (pon-james) [reply pull-requests.md:21]: Not exactly sure what you expect here.

File: pull-requests.md [hunk 16]
```diff
@@ -0,0 +1,162 @@
...
+* It will have a title that reflects the work within, and a summary that helps to understand the context of the change.
+* There will be well written commit messages, with well crafted commits that tell the story of the development of this work.
+* Ideally it will be small and easy to understand. Single commit PRs are usually easy to submit, review, and merge.
+* The code contained within will meet the [best practices](./README.md#best-practices) wherever possible.
```
[the remark this answers is not in the document]
Comment (pon-james) [reply pull-requests.md]: All points are superflous? Please elaborate.

**Pull request merged**